### DISCUSSION TITLE
How can i get sol address from privateKey:(Nno9dBF53drmKNYbaXNKxyAf5KMFfMzudVXQgLcBgpzWqMJ5dK6MfsR51gaDPhsNLgEj2hkStnqGFU7iuBFmWF2)

### DISCUSSION THREAD

**Discussion by deeryoona** [trustwallet/wallet-core#3815] [Q&A]
i am an iOS developer
 if let prik = PrivateKey(data: Data(hexString: "Nno9dBF53drmKNYbaXNKxyAf5KMFfMzudVXQgLcBgpzWqMJ5dK6MfsR51gaDPhsNLgEj2hkStnqGFU7iuBFmWF2")!) {
                  let address = CoinType.solana.deriveAddress(privateKey: prik)
                  return address
              }else{
                  return ""
              }

i can not get sol address,i do not know how to get the correct sol address

**Comment by satoshiotomakan** [comment]
Hi @deeryoona, please create a discussion, tickets are for bugs/improvements only.

Could you please specify which Solana address you expect to get? Did you export this private key from an external wallet?

**Reply by deeryoona** [reply]
i want get address:5War5zZsJDUpXRLC2d2SKdvzfponUk1NzPizKe3ZfGn4
this private key from Phantom
and this privateKey can import TrustWallet App

**Reply by satoshiotomakan** [reply]
First of all, "Nno9dBF53drmKNYbaXNKxyAf5KMFfMzudVXQgLcBgpzWqMJ5dK6MfsR51gaDPhsNLgEj2hkStnqGFU7iuBFmWF2" is not a hex, but base58.
You should decode it as base58, and create a `PrivateKey` with the data:
(note I haven't tested the code, so there can be minor mistakes)

```swift
let secret = Base58.decode(string="Nno9dBF53drmKNYbaXNKxyAf5KMFfMzudVXQgLcBgpzWqMJ5dK6MfsR51gaDPhsNLgEj2hkStnqGFU7iuBFmWF2")
let privateKey = PrivateKey(secret)
let derivedAddress = CoinType.solana.deriveAddress(privateKey: privateKey)
```

**Comment by deeryoona** [comment]
> ```swift
> let secret = Base58.decode(string="Nno9dBF53drmKNYbaXNKxyAf5KMFfMzudVXQgLcBgpzWqMJ5dK6MfsR51gaDPhsNLgEj2hkStnqGFU7iuBFmWF2")
> let privateKey = PrivateKey(secret)
> ```
like this :
 func getSolOriginAddressWithPriKey(priKey: String) -> String{
        let privateKeyData = Data(Base58.decodeNoCheck(string: priKey)!)
        
        let derivedAddress = CoinType.solana.deriveAddress(privateKey: PrivateKey(data: privateKeyData) ?? PrivateKey())
        
        return derivedAddress
    }
the getted address is:5WLG3i5cmJTobKyzTkF1gkyH7nteV1ArMh54vU4WXBCt
Not :5War5zZsJDUpXRLC2d2SKdvzfponUk1NzPizKe3ZfGn4
i needed address is 5War5zZsJDUpXRLC2d2SKdvzfponUk1NzPizKe3ZfGn4,like trustwallet App
![1991713945605_ pic](https://github.com/trustwallet/wallet-core/assets/5423076/2b2b2f4f-b467-4f6a-9c51-e0d143e62731)

**Comment by RomanSytnyk** [comment]
I've found a solution for that issue.
A Solana private key consists of 32 bytes of seed and 32 bytes of public key.
If you pass the full private key (64 bytes) into `PrivateKey(data: keyData)`, it will return `nil` (in Swift for iOS).

You need to pass only the first 32 bytes to get the private key with this SDK:
`PrivateKey(data: keyData.prefix(32))`

This will return a working PrivateKey.

@deeryoona I think it would help you

**Comment by Lawan012** [comment]
[Jobjhcjvgi]()

**Comment by Lawan012** [comment]
[Jobjhcjvgi]()
